### PR TITLE
Feature/cheby

### PR DESCRIPTION
**Pull request by pohm01**
Add a Chebyshev iteration as a linear solver.

Added as a ksp_t since it fit that data structure well enough (despite not being a Krylov method).

### CODE REVIEW

File: src/krylov/bcknd/cpu/cheby.f90 [hunk 1]
```diff
@@ -0,0 +1,252 @@
+! Copyright (c) 2020-2021, The Neko Authors
```
Comment (njansson) [inline src/krylov/bcknd/cpu/cheby.f90:1]:
```suggestion
! Copyright (c) 2024, The Neko Authors
```

File: src/.depends [hunk 2]
```diff
@@ -147,7 +147,7 @@ krylov/precon.o : krylov/precon.f90 config/num_types.o
...
+krylov/krylov_fctry.o : krylov/krylov_fctry.f90 config/neko_config.o common/utils.o krylov/precon.o krylov/krylov.o config/num_types.o krylov/bcknd/device/gmres_device.o krylov/bcknd/sx/gmres_sx.o krylov/bcknd/cpu/gmres.o krylov/bcknd/cpu/bicgstab.o krylov/bcknd/device/fusedcg_cpld_device.o krylov/bcknd/device/fusedcg_device.o krylov/bcknd/device/pipecg_device.o krylov/bcknd/sx/pipecg_sx.o krylov/bcknd/cpu/pipecg.o krylov/bcknd/cpu/cacg.o krylov/bcknd/device/cg_device.o krylov/bcknd/cpu/cg_coupled.o krylov/bcknd/sx/cg_sx.o krylov/bcknd/cpu/cg.o krylov/bcknd/cpu/cheby.o
 krylov/bcknd/cpu/cg.o : krylov/bcknd/cpu/cg.f90 comm/comm.o math/math.o bc/bc.o gs/gather_scatter.o sem/coef.o field/field.o math/ax.o krylov/precon.o krylov/krylov.o config/num_types.o 
 krylov/bcknd/cpu/cacg.o : krylov/bcknd/cpu/cacg.f90 math/mxm_wrapper.o comm/comm.o common/utils.o math/math.o bc/bc.o gs/gather_scatter.o sem/coef.o field/field.o math/ax.o krylov/precon.o krylov/krylov.o config/num_types.o 
 krylov/bcknd/cpu/pipecg.o : krylov/bcknd/cpu/pipecg.f90 comm/comm.o math/math.o bc/bc.o gs/gather_scatter.o sem/coef.o field/field.o config/num_types.o math/ax.o krylov/precon.o krylov/krylov.o
```
Comment (njansson) [inline src/.depends:153]:
```suggestion
krylov/bcknd/cpu/cheby.o : krylov/bcknd/cpu/cheby.f90 comm/comm.o math/math.o bc/bc.o gs/gather_scatter.o sem/space.o mesh/mesh.o sem/coef.o field/field.o config/num_types.o math/ax.o krylov/precon.o krylov/krylov.o
krylov/bcknd/cpu/pipecg.o : krylov/bcknd/cpu/pipecg.f90 comm/comm.o math/math.o bc/bc.o gs/gather_scatter.o sem/coef.o field/field.o config/num_types.o math/ax.o krylov/precon.o krylov/krylov.o 
```

File: src/krylov/bcknd/cpu/cheby.f90 [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+
+    if (present(rel_tol) .and. present(abs_tol)) then
+       call this%ksp_init(max_iter, rel_tol, abs_tol)
+    else if (present(rel_tol)) then
+       call this%ksp_init(max_iter, rel_tol=rel_tol)
+    else if (present(abs_tol)) then
+       call this%ksp_init(max_iter, abs_tol=abs_tol)
+    else
+       call this%ksp_init(max_iter)
+    end if
```
Comment (njansson) [inline src/krylov/bcknd/cpu/cheby.f90:94]:
```suggestion
    if (present(rel_tol) .and. present(abs_tol)) then
       call this%ksp_init(max_iter, rel_tol, abs_tol)
    else if (present(rel_tol)) then
       call this%ksp_init(max_iter, rel_tol = rel_tol)
    else if (present(abs_tol)) then
       call this%ksp_init(max_iter, abs_tol = abs_tol)
    else
       call this%ksp_init(max_iter)
    end if
```

**Review by njansson — APPROVED** [review]
🍻  Indeed we should stash this somewhere else, maybe have a `linear_solver_t` that can be either `iterative_t` (w. `ksp_t` and cheby) and a `multilevel_t` for multigrid kind of solvers

File: src/krylov/bcknd/cpu/cheby.f90 [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+  use precon, only : pc_t
+  use ax_product, only : ax_t
+  use num_types, only: rp
+  use field, only : field_t
+  use coefs, only : coef_t
+  use mesh, only : mesh_t
+  use space, only : space_t
+  use gather_scatter, only : gs_t, GS_OP_ADD
+  use bc, only : bc_list_t, bc_list_apply
+  use math, only : glsc3, rzero, rone, copy, sub2, cmult2, abscmp, glsc2, add2s1, add2s2
```
Comment (njansson) [inline src/krylov/bcknd/cpu/cheby.f90:45]:
```suggestion
  use math, only : glsc3, rzero, rone, copy, sub2, cmult2, abscmp, glsc2,  &
       add2s1, add2s2
```

File: src/krylov/bcknd/cpu/cheby.f90 [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+      a = lam / lam_factor
+      this%tha = (b+a)/2.0_rp
+      this%dlt = (b-a)/2.0_rp
+
+      this%recompute_eigs = .false.
+    end associate
+  end subroutine cheby_power
+
+  !> A chebyshev preconditioner
+  function cheby_solve(this, Ax, x, f, n, coef, blst, gs_h, niter) result(ksp_results)
```
Comment (njansson) [inline src/krylov/bcknd/cpu/cheby.f90:156]:
```suggestion
  function cheby_solve(this, Ax, x, f, n, coef, blst, gs_h, niter) &
       result(ksp_results)
```

**Review by njansson — APPROVED** [review]

File: src/.depends [hunk 6]
```diff
@@ -147,9 +147,10 @@ krylov/precon.o : krylov/precon.f90 config/num_types.o
...
-krylov/krylov_fctry.o : krylov/krylov_fctry.f90 config/neko_config.o common/utils.o krylov/precon.o krylov/krylov.o config/num_types.o krylov/bcknd/device/gmres_device.o krylov/bcknd/sx/gmres_sx.o krylov/bcknd/cpu/gmres.o krylov/bcknd/cpu/bicgstab.o krylov/bcknd/device/fusedcg_cpld_device.o krylov/bcknd/device/fusedcg_device.o krylov/bcknd/device/pipecg_device.o krylov/bcknd/sx/pipecg_sx.o krylov/bcknd/cpu/pipecg.o krylov/bcknd/cpu/cacg.o krylov/bcknd/device/cg_device.o krylov/bcknd/cpu/cg_coupled.o krylov/bcknd/sx/cg_sx.o krylov/bcknd/cpu/cg.o 
+krylov/krylov_fctry.o : krylov/krylov_fctry.f90 config/neko_config.o common/utils.o krylov/precon.o krylov/krylov.o config/num_types.o krylov/bcknd/device/gmres_device.o krylov/bcknd/sx/gmres_sx.o krylov/bcknd/cpu/gmres.o krylov/bcknd/cpu/bicgstab.o krylov/bcknd/device/fusedcg_cpld_device.o krylov/bcknd/device/fusedcg_device.o krylov/bcknd/device/pipecg_device.o krylov/bcknd/sx/pipecg_sx.o krylov/bcknd/cpu/pipecg.o krylov/bcknd/cpu/cacg.o krylov/bcknd/device/cg_device.o krylov/bcknd/cpu/cg_coupled.o krylov/bcknd/sx/cg_sx.o krylov/bcknd/cpu/cg.o krylov/bcknd/cpu/cheby.o
```
Comment (njansson) [inline src/.depends:150]:
```suggestion
krylov/krylov_fctry.o : krylov/krylov_fctry.f90 config/neko_config.o common/utils.o krylov/precon.o krylov/krylov.o config/num_types.o krylov/bcknd/device/gmres_device.o krylov/bcknd/sx/gmres_sx.o krylov/bcknd/cpu/gmres.o krylov/bcknd/cpu/bicgstab.o krylov/bcknd/device/fusedcg_cpld_device.o krylov/bcknd/device/fusedcg_device.o krylov/bcknd/device/pipecg_device.o krylov/bcknd/sx/pipecg_sx.o krylov/bcknd/cpu/pipecg.o krylov/bcknd/cpu/cacg.o krylov/bcknd/device/cg_device.o krylov/bcknd/cpu/cg_coupled.o krylov/bcknd/sx/cg_sx.o krylov/bcknd/cpu/cg.o krylov/bcknd/cpu/cheby.o 
```

File: src/.depends [hunk 7]
```diff
@@ -147,9 +147,10 @@ krylov/precon.o : krylov/precon.f90 config/num_types.o
...
 krylov/bcknd/cpu/cg.o : krylov/bcknd/cpu/cg.f90 comm/comm.o math/math.o bc/bc.o gs/gather_scatter.o sem/coef.o field/field.o math/ax.o krylov/precon.o krylov/krylov.o config/num_types.o 
 krylov/bcknd/cpu/cacg.o : krylov/bcknd/cpu/cacg.f90 math/mxm_wrapper.o comm/comm.o common/utils.o math/math.o bc/bc.o gs/gather_scatter.o sem/coef.o field/field.o math/ax.o krylov/precon.o krylov/krylov.o config/num_types.o 
+krylov/bcknd/cpu/cheby.o : krylov/bcknd/cpu/cheby.f90 comm/comm.o math/math.o bc/bc.o gs/gather_scatter.o sem/space.o mesh/mesh.o sem/coef.o field/field.o config/num_types.o math/ax.o krylov/precon.o krylov/krylov.o
```
Comment (njansson) [inline src/.depends:153]:
```suggestion
krylov/bcknd/cpu/cheby.o : krylov/bcknd/cpu/cheby.f90 comm/comm.o math/math.o bc/bc.o gs/gather_scatter.o sem/space.o mesh/mesh.o sem/coef.o field/field.o config/num_types.o math/ax.o krylov/precon.o krylov/krylov.o 
```

**Review by njansson — APPROVED** [review]

File: src/.depends [hunk 8]
```diff
@@ -147,9 +147,10 @@ krylov/precon.o : krylov/precon.f90 config/num_types.o
...
-krylov/krylov_fctry.o : krylov/krylov_fctry.f90 config/neko_config.o common/utils.o krylov/precon.o krylov/krylov.o config/num_types.o krylov/bcknd/device/gmres_device.o krylov/bcknd/sx/gmres_sx.o krylov/bcknd/cpu/gmres.o krylov/bcknd/cpu/bicgstab.o krylov/bcknd/device/fusedcg_cpld_device.o krylov/bcknd/device/fusedcg_device.o krylov/bcknd/device/pipecg_device.o krylov/bcknd/sx/pipecg_sx.o krylov/bcknd/cpu/pipecg.o krylov/bcknd/cpu/cacg.o krylov/bcknd/device/cg_device.o krylov/bcknd/cpu/cg_coupled.o krylov/bcknd/sx/cg_sx.o krylov/bcknd/cpu/cg.o 
+krylov/krylov_fctry.o : krylov/krylov_fctry.f90 config/neko_config.o common/utils.o krylov/precon.o krylov/krylov.o config/num_types.o krylov/bcknd/device/gmres_device.o krylov/bcknd/sx/gmres_sx.o krylov/bcknd/cpu/gmres.o krylov/bcknd/cpu/bicgstab.o krylov/bcknd/device/fusedcg_cpld_device.o krylov/bcknd/device/fusedcg_device.o krylov/bcknd/device/pipecg_device.o krylov/bcknd/sx/pipecg_sx.o krylov/bcknd/cpu/pipecg.o krylov/bcknd/cpu/cacg.o krylov/bcknd/device/cg_device.o krylov/bcknd/cpu/cg_coupled.o krylov/bcknd/sx/cg_sx.o krylov/bcknd/cpu/cg.o krylov/bcknd/cpu/cheby.o
```
Comment (njansson) [inline src/.depends:150]:
```suggestion
krylov/krylov_fctry.o : krylov/krylov_fctry.f90 config/neko_config.o common/utils.o krylov/precon.o krylov/krylov.o config/num_types.o krylov/bcknd/device/gmres_device.o krylov/bcknd/sx/gmres_sx.o krylov/bcknd/cpu/cheby.o krylov/bcknd/cpu/gmres.o krylov/bcknd/cpu/bicgstab.o krylov/bcknd/device/fusedcg_cpld_device.o krylov/bcknd/device/fusedcg_device.o krylov/bcknd/device/pipecg_device.o krylov/bcknd/sx/pipecg_sx.o krylov/bcknd/cpu/pipecg.o krylov/bcknd/cpu/cacg.o krylov/bcknd/device/cg_device.o krylov/bcknd/cpu/cg_coupled.o krylov/bcknd/sx/cg_sx.o krylov/bcknd/cpu/cg.o 
```

**Review by njansson — APPROVED** [review]

File: src/krylov/bcknd/cpu/cheby.f90 [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+  use bc, only : bc_list_t, bc_list_apply
+  use math, only : glsc3, rzero, rone, copy, sub2, cmult2, abscmp, glsc2,  &
+       add2s1, add2s2
+  use comm
+  implicit none
+  private
+
+  !> Defines a Chebyshev preconditioner
+  type, public, extends(ksp_t) :: cheby_t
+     !!!!!!class(ax_t), pointer :: ax
```
Comment (njansson) [inline src/krylov/bcknd/cpu/cheby.f90:53]:
```suggestion
```

**Review by njansson — APPROVED** [review]

**Review by MaansAndersson — APPROVED** [review]

**Pull request merged**